Design/histogram_type.md

```diff
@@ -0,0 +1,108 @@
+
+The problems include:
+
+- No type safety:
+  - The bug in `ConvertUnits` creped in because we are currently using the same instance of a single type for both histogram data and distribution data. The actual meaning of a vector of `Y` values is basically kept as a mental flag in the brain of a Mantid developer.
+  - We are using a single type for `X`, `Y`, and `E`. It is thus easy to accidentally mix them up.
+  - The array `E` of error values is used for both the variance and the standard deviation. The state is again carried as a flag in the brain of a developer. We even have functions such as `rebin` that return an array of `E` values that may mean either variance or standard deviation depending on a flag to the function!
+- Code duplication: Re-implementing the same thing 10 times is wasteful and error prone.
+  - Operations like adding histograms are frequent.
+  - Propagation of errors.
+  - Conversions between variance and standard deviation.
+- Unit testing: Implementation of operations on histogram are currently mixed into algorithm code. Unit tests thus need to test two things at the same time.
+  - Many units tests of algorithms do not fully test the range of errors that can happen when, e.g., adding two spectra. Doing so would be very hard in the current de-centralized implementation,
+  - Many algorithms need to modify the errors. In some algorithms the unit tests cover the errors, in others they do not.
+
+
+## Design goals
+
+- Provide simple arithmetic operations for histograms.
+  - Arithmetic operations automatically deal with propagation of errors.
+- Hide differences between histogram and distribution data.
+  - Internal storage type could be either, does not need to be known at the interface level?
+  - Arithmetic operations should do the right thing for any data type automatically.
+- Hide error handling.
+  - For simple operations with histograms algorithm authors should not need to know how to propagate errors.
+- Enable unit testing of key concepts in Mantid.
+  - Arithmetic operations with histograms.
+  - Propagation of errors.
+
+
+
+## Implementation
+
+#### Basics
+
+The details of the implementation have not been worked out yet.
+The basic concept seems simple enough:
+
+- Add one class for a histogram with an interface that provides standard operations for histogram data.
+  - Note that we do not introduce a separate type for histogram and distribution data. This distinction would be only internal, and access to either histogram or distribution data is provided via public interface methods. The goal is to eliminate this distinction from as much of the code as possible.
+- Add a few classes for `X`, `Y`, and `E` (etc.) that would mainly be used internally in the histogram class and for direct access to the underlying data. If these inherit from `std::vetor<double>`, they could also be used to keep the legacy interface for `ISpectrum` alive until the roll-out is complete.
+
```

Comment:
@martyngigg I agree that is not ideal (in fact I read that exact stackoverflow comment before writing this design document).

I think there are a few things to consider when taking a decision:

1. Compatibility with the current interface: This is mainly a roll-out issue, but it may take a long time until it is complete, since `X`, `Y`, and `E` are used everywhere. Can we achieve the same thing if we let, e.g, `readX` return a reference to a `std::vector<double>` *member* of a new histogram type?
2. We do not need polymorphism. When inheriting that would still leave the risk of deleting via a base class pointer. Do we generally avoid that?
3. If we inherit from `std::vector`, we would still implement most functionality via algorithms.
4. Would inheriting from `std::vector` vialotate the Liskov substitution principle? I am tempted to say so, which would be good, but I am not entirely sure.
5. `std::vector` has quite a big and nice interface (e.g., for initialization), which is used frequently. If we have the vector as a member we need to replicate that for each of our new types, which adds a lot of code.
6. Performance: If we chose composition have have one extra indirection for data access. If we inline those access methods (e.g., `operator[]`) there should be little to no negative performance impact.

I think writing this clarified a few things also for myself. If we can find a nice way of dealing with 1.) the only remaining reason for inheritance is 5.). 5.) is a rather 'local' problem, so it is dealt with easily with some extra work.